File: bindings/dotnet/src/Wcl/Wasm/WasmRuntime.cs
using System;
using System.IO;
using System.Text;
using Wasmtime;

namespace Wcl.Wasm
{
    internal sealed class WasmRuntime : IDisposable
    {
        private static readonly Lazy<WasmRuntime> _instance = new Lazy<WasmRuntime>(() => new WasmRuntime());
        internal static WasmRuntime Instance => _instance.Value;

        private readonly Engine _engine;
        private readonly Module _module;
        private readonly Linker _linker;
        private Store _store;
        private Instance? _instance_wasm;
        private readonly object _lock = new object();

        // Cached function references
        private Func<int, int>? _alloc;
        private Action<int, int>? _dealloc;
        private Func<int, int, int>? _parse;
        private Func<int, int, int, int>? _parseWithFunctions;
        private Action<int>? _documentFree;
        private Func<int, int>? _documentValues;
        private Func<int, int>? _documentHasErrors;
        private Func<int, int>? _documentDiagnostics;
        private Func<int, int, int>? _documentQuery;
        private Func<int, int>? _documentBlocks;
        private Func<int, int, int>? _documentBlocksOfType;
        private Action<int>? _stringFree;

        private WasmRuntime()
        {
            _engine = new Engine();
            var wasmBytes = LoadEmbeddedWasm();
            _module = Module.FromBytes(_engine, "wcl_wasm", wasmBytes);
            _linker = new Linker(_engine);
            _linker.DefineWasi();
            DefineHostFunctions();
            _store = new Store(_engine);
            _store.SetWasiConfiguration(new WasiConfiguration());
            _instance_wasm = _linker.Instantiate(_store, _module);
            CacheFunctions();
        }

        private static byte[] LoadEmbeddedWasm()
        {
            var assembly = typeof(WasmRuntime).Assembly;
            var resourceName = "Wcl.Wasm.wcl_wasm.wasm";

            using var stream = assembly.GetManifestResourceStream(resourceName);
            if (stream == null)
            {
                // Fall back to checking all resource names
                var names = assembly.GetManifestResourceNames();
                foreach (var name in names)
                {
                    if (name.EndsWith("wcl_wasm.wasm"))
                    {
                        using var s = assembly.GetManifestResourceStream(name);
                        if (s != null)
                        {
                            using var ms2 = new MemoryStream();
                            s.CopyTo(ms2);
                            return ms2.ToArray();
                        }
                    }
                }
                throw new FileNotFoundException(
                    $"Embedded WASM resource not found. Available: [{string.Join(", ", names)}]");
            }

            using var ms = new MemoryStream();
            stream.CopyTo(ms);
            return ms.ToArray();
        }

        private void DefineHostFunctions()
        {
            _linker.DefineFunction("env", "host_call_function",
                (Caller caller, int namePtr, int nameLen, int argsPtr, int argsLen, int resultPtrOut, int resultLenOut) =>
                {
                    var memory = caller.GetMemory("memory");
                    if (memory == null) return -1;

                    var nameBytes = memory.GetSpan(namePtr, nameLen);
                    var name = Encoding.UTF8.GetString(nameBytes);

                    var argsBytes = memory.GetSpan(argsPtr, argsLen);
                    var argsJson = Encoding.UTF8.GetString(argsBytes);

                    var (success, resultJson) = WasmCallbackBridge.Invoke(name, argsJson);

                    if (resultJson != null)
                    {
                        var resultBytes = Encoding.UTF8.GetBytes(resultJson);
                        var allocFn = caller.GetFunction("wcl_wasm_alloc");
                        if (allocFn == null) return -1;
                        var ptr = (int)allocFn.Invoke(resultBytes.Length)!;

                        var dest = memory.GetSpan(ptr, resultBytes.Length);
                        resultBytes.CopyTo(dest);

                        // Write pointer and length to output params
                        var memSpan = memory.GetSpan<byte>(0, (int)memory.GetLength());
                        BitConverter.TryWriteBytes(memSpan.Slice(resultPtrOut), ptr);
                        BitConverter.TryWriteBytes(memSpan.Slice(resultLenOut), resultBytes.Length);
                    }

                    return success ? 0 : -1;
                });
        }

        private void CacheFunctions()
        {
            _alloc = _instance_wasm!.GetFunction<int, int>("wcl_wasm_alloc");
            _dealloc = _instance_wasm.GetAction<int, int>("wcl_wasm_dealloc");
            _parse = _instance_wasm.GetFunction<int, int, int>("wcl_wasm_parse");
            _parseWithFunctions = _instance_wasm.GetFunction<int, int, int, int>("wcl_wasm_parse_with_functions");
            _documentFree = _instance_wasm.GetAction<int>("wcl_wasm_document_free");
            _documentValues = _instance_wasm.GetFunction<int, int>("wcl_wasm_document_values");
            _documentHasErrors = _instance_wasm.GetFunction<int, int>("wcl_wasm_document_has_errors");
            _documentDiagnostics = _instance_wasm.GetFunction<int, int>("wcl_wasm_document_diagnostics");
            _documentQuery = _instance_wasm.GetFunction<int, int, int>("wcl_wasm_document_query");
            _documentBlocks = _instance_wasm.GetFunction<int, int>("wcl_wasm_document_blocks");
            _documentBlocksOfType = _instance_wasm.GetFunction<int, int, int>("wcl_wasm_document_blocks_of_type");
            _stringFree = _instance_wasm.GetAction<int>("wcl_wasm_string_free");
        }

        private Memory GetMemory()
        {
            return _instance_wasm!.GetMemory("memory")
                ?? throw new InvalidOperationException("WASM memory not found");
        }

        internal int WriteString(string? s)
        {
            if (s == null) return 0;
            var bytes = Encoding.UTF8.GetBytes(s);
            var ptr = _alloc!(bytes.Length + 1);
            var memory = GetMemory();
            var dest = memory.GetSpan(ptr, bytes.Length + 1);
            bytes.CopyTo(dest);
            dest[bytes.Length] = 0; // null terminator
            return ptr;
        }

        internal string ReadCString(int ptr)
        {
            if (ptr == 0) return "";
            var memory = GetMemory();
            var span = memory.GetSpan<byte>(0, (int)memory.GetLength());
            int start = ptr;
            int end = start;
            while (end < span.Length && span[end] != 0) end++;
            return Encoding.UTF8.GetString(span.Slice(start, end - start));
        }

        internal string ConsumeString(int ptr)
        {
            if (ptr == 0) return "";
            var s = ReadCString(ptr);
            _stringFree!(ptr);
            return s;
        }

        internal void FreeWasmString(int ptr)
        {
            if (ptr != 0) _stringFree!(ptr);
        }

        // ── Public API ──────────────────────────────────────────────────────

        internal int Parse(string source, string? optionsJson)
        {
            lock (_lock)
            {
                var srcPtr = WriteString(source);
                var optsPtr = WriteString(optionsJson);
                try
                {
                    return _parse!(srcPtr, optsPtr);
                }
                finally
                {
                    if (srcPtr != 0) _dealloc!(srcPtr, Encoding.UTF8.GetByteCount(source) + 1);
                    if (optsPtr != 0 && optionsJson != null) _dealloc!(optsPtr, Encoding.UTF8.GetByteCount(optionsJson) + 1);
                }
            }
        }

        internal int ParseWithFunctions(string source, string? optionsJson, string funcNamesJson)
        {
            lock (_lock)
            {
                var srcPtr = WriteString(source);
                var optsPtr = WriteString(optionsJson);
                var namesPtr = WriteString(funcNamesJson);
                try
                {
                    return _parseWithFunctions!(srcPtr, optsPtr, namesPtr);
                }
                finally
                {
                    if (srcPtr != 0) _dealloc!(srcPtr, Encoding.UTF8.GetByteCount(source) + 1);
                    if (optsPtr != 0 && optionsJson != null) _dealloc!(optsPtr, Encoding.UTF8.GetByteCount(optionsJson) + 1);
                    if (namesPtr != 0) _dealloc!(namesPtr, Encoding.UTF8.GetByteCount(funcNamesJson) + 1);
                }
            }
        }

        internal void DocumentFree(int handle)
        {
            lock (_lock) { _documentFree!(handle); }
        }

        internal string DocumentValues(int handle)
        {
            lock (_lock)
            {
                var ptr = _documentValues!(handle);
                return ConsumeString(ptr);
            }
        }

        internal bool DocumentHasErrors(int handle)
        {
            lock (_lock) { return _documentHasErrors!(handle) != 0; }
        }

        internal string DocumentDiagnostics(int handle)
        {
            lock (_lock)
            {
                var ptr = _documentDiagnostics!(handle);
                return ConsumeString(ptr);
            }
        }

        internal string DocumentQuery(int handle, string query)
        {
            lock (_lock)
            {
                var qPtr = WriteString(query);
                try
                {
                    var ptr = _documentQuery!(handle, qPtr);
                    return ConsumeString(ptr);
                }
                finally
                {
                    if (qPtr != 0) _dealloc!(qPtr, Encoding.UTF8.GetByteCount(query) + 1);
                }
            }
        }

        internal string DocumentBlocks(int handle)
        {
            lock (_lock)
            {
                var ptr = _documentBlocks!(handle);
                return ConsumeString(ptr);
            }
        }

        internal string DocumentBlocksOfType(int handle, string kind)
        {
            lock (_lock)
            {
                var kPtr = WriteString(kind);
                try
                {
                    var ptr = _documentBlocksOfType!(handle, kPtr);
                    return ConsumeString(ptr);
                }
                finally
                {
                    if (kPtr != 0) _dealloc!(kPtr, Encoding.UTF8.GetByteCount(kind) + 1);
                }
            }
        }

        public void Dispose()
        {
            _instance_wasm = null;
            _store?.Dispose();
            _module?.Dispose();
            _engine?.Dispose();
        }
    }
}

File: bindings/jvm/src/main/java/io/github/wiltaylor/wcl/wasm/WasmRuntime.java
package io.github.wiltaylor.wcl.wasm;

import com.dylibso.chicory.runtime.ExportFunction;
import com.dylibso.chicory.runtime.HostFunction;
import com.dylibso.chicory.runtime.ImportFunction;
import com.dylibso.chicory.runtime.ImportValues;
import com.dylibso.chicory.runtime.Instance;
import com.dylibso.chicory.runtime.Memory;
import com.dylibso.chicory.wasm.Parser;
import com.dylibso.chicory.wasm.types.ValueType;
import com.dylibso.chicory.wasi.WasiOptions;
import com.dylibso.chicory.wasi.WasiPreview1;
import io.github.wiltaylor.wcl.WclException;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class WasmRuntime {
    private static final WasmRuntime INSTANCE = new WasmRuntime();

    public static WasmRuntime getInstance() { return INSTANCE; }

    private final Instance instance;
    private final Object lock = new Object();

    // Cached export references
    private final ExportFunction alloc;
    private final ExportFunction dealloc;
    private final ExportFunction parse;
    private final ExportFunction parseWithFunctions;
    private final ExportFunction documentFree;
    private final ExportFunction documentValues;
    private final ExportFunction documentHasErrors;
    private final ExportFunction documentDiagnostics;
    private final ExportFunction documentQuery;
    private final ExportFunction documentBlocks;
    private final ExportFunction documentBlocksOfType;
    private final ExportFunction stringFree;

    private WasmRuntime() {
        try {
            var wasmBytes = loadEmbeddedWasm();
            var module = Parser.parse(wasmBytes);

            var wasi = WasiPreview1.builder()
                    .withOptions(WasiOptions.builder().build())
                    .build();

            var hostCallFunction = new HostFunction(
                    "env",
                    "host_call_function",
                    List.of(ValueType.I32, ValueType.I32, ValueType.I32,
                            ValueType.I32, ValueType.I32, ValueType.I32),
                    List.of(ValueType.I32),
                    (inst, args) -> {
                        var memory = inst.memory();
                        int namePtr = (int) args[0];
                        int nameLen = (int) args[1];
                        int argsPtr = (int) args[2];
                        int argsLen = (int) args[3];
                        int resultPtrOut = (int) args[4];
                        int resultLenOut = (int) args[5];

                        var nameBytes = memory.readBytes(namePtr, nameLen);
                        var name = new String(nameBytes, StandardCharsets.UTF_8);

                        var argsBytes = memory.readBytes(argsPtr, argsLen);
                        var argsJson = new String(argsBytes, StandardCharsets.UTF_8);

                        var result = WasmCallbackBridge.invoke(name, argsJson);

                        if (result.resultJson() != null) {
                            var resultBytes = result.resultJson().getBytes(StandardCharsets.UTF_8);
                            var allocFn = inst.export("wcl_wasm_alloc");
                            var ptr = (int) allocFn.apply(resultBytes.length)[0];

                            memory.write(ptr, resultBytes);
                            memory.writeI32(resultPtrOut, ptr);
                            memory.writeI32(resultLenOut, resultBytes.length);
                        }

                        return new long[] { result.success() ? 0 : -1 };
                    }
            );

            // Combine WASI host functions with our custom host function
            var wasiHostFunctions = wasi.toHostFunctions();
            var allFunctions = new ArrayList<ImportFunction>(Arrays.asList(wasiHostFunctions));
            allFunctions.add(hostCallFunction);

            var imports = ImportValues.builder()
                    .withFunctions(allFunctions)
                    .build();

            instance = Instance.builder(module)
                    .withImportValues(imports)
                    .build();

            alloc = instance.export("wcl_wasm_alloc");
            dealloc = instance.export("wcl_wasm_dealloc");
            parse = instance.export("wcl_wasm_parse");
            parseWithFunctions = instance.export("wcl_wasm_parse_with_functions");
            documentFree = instance.export("wcl_wasm_document_free");
            documentValues = instance.export("wcl_wasm_document_values");
            documentHasErrors = instance.export("wcl_wasm_document_has_errors");
            documentDiagnostics = instance.export("wcl_wasm_document_diagnostics");
            documentQuery = instance.export("wcl_wasm_document_query");
            documentBlocks = instance.export("wcl_wasm_document_blocks");
            documentBlocksOfType = instance.export("wcl_wasm_document_blocks_of_type");
            stringFree = instance.export("wcl_wasm_string_free");

        } catch (IOException e) {
            throw new WclException("failed to load WASM module", e);
        }
    }

    private static byte[] loadEmbeddedWasm() throws IOException {
        var stream = WasmRuntime.class.getResourceAsStream(
                "/io/github/wiltaylor/wcl/wasm/wcl_wasm.wasm");
        if (stream == null) {
            throw new IOException("embedded WASM resource not found");
        }
        try (stream) {
            return stream.readAllBytes();
        }
    }

    private Memory getMemory() {
        return instance.memory();
    }

    int writeString(String s) {
        if (s == null) return 0;
        var bytes = s.getBytes(StandardCharsets.UTF_8);
        var ptr = (int) alloc.apply(bytes.length + 1)[0];
        var memory = getMemory();
        memory.write(ptr, bytes);
        memory.write(ptr + bytes.length, new byte[] { 0 }, 0, 1); // null terminator
        return ptr;
    }

    String readCString(int ptr) {
        if (ptr == 0) return "";
        var memory = getMemory();
        int end = ptr;
        while (memory.read(end) != 0) end++;
        var bytes = memory.readBytes(ptr, end - ptr);
        return new String(bytes, StandardCharsets.UTF_8);
    }

    String consumeString(int ptr) {
        if (ptr == 0) return "";
        var s = readCString(ptr);
        stringFree.apply(ptr);
        return s;
    }

    // Public API

    public int parse(String source, String optionsJson) {
        synchronized (lock) {
            var srcPtr = writeString(source);
            var optsPtr = writeString(optionsJson);
            try {
                return (int) parse.apply(srcPtr, optsPtr)[0];
            } finally {
                if (srcPtr != 0) dealloc.apply(srcPtr, source.getBytes(StandardCharsets.UTF_8).length + 1);
                if (optsPtr != 0 && optionsJson != null)
                    dealloc.apply(optsPtr, optionsJson.getBytes(StandardCharsets.UTF_8).length + 1);
            }
        }
    }

    public int parseWithFunctions(String source, String optionsJson, String funcNamesJson) {
        synchronized (lock) {
            var srcPtr = writeString(source);
            var optsPtr = writeString(optionsJson);
            var namesPtr = writeString(funcNamesJson);
            try {
                return (int) parseWithFunctions.apply(srcPtr, optsPtr, namesPtr)[0];
            } finally {
                if (srcPtr != 0) dealloc.apply(srcPtr, source.getBytes(StandardCharsets.UTF_8).length + 1);
                if (optsPtr != 0 && optionsJson != null)
                    dealloc.apply(optsPtr, optionsJson.getBytes(StandardCharsets.UTF_8).length + 1);
                if (namesPtr != 0)
                    dealloc.apply(namesPtr, funcNamesJson.getBytes(StandardCharsets.UTF_8).length + 1);
            }
        }
    }

    public void documentFree(int handle) {
        synchronized (lock) { documentFree.apply(handle); }
    }

    public String documentValues(int handle) {
        synchronized (lock) {
            var ptr = (int) documentValues.apply(handle)[0];
            return consumeString(ptr);
        }
    }

    public boolean documentHasErrors(int handle) {
        synchronized (lock) { return documentHasErrors.apply(handle)[0] != 0; }
    }

    public String documentDiagnostics(int handle) {
        synchronized (lock) {
            var ptr = (int) documentDiagnostics.apply(handle)[0];
            return consumeString(ptr);
        }
    }

    public String documentQuery(int handle, String query) {
        synchronized (lock) {
            var qPtr = writeString(query);
            try {
                var ptr = (int) documentQuery.apply(handle, qPtr)[0];
                return consumeString(ptr);
            } finally {
                if (qPtr != 0) dealloc.apply(qPtr, query.getBytes(StandardCharsets.UTF_8).length + 1);
            }
        }
    }

    public String documentBlocks(int handle) {
        synchronized (lock) {
            var ptr = (int) documentBlocks.apply(handle)[0];
            return consumeString(ptr);
        }
    }

    public String documentBlocksOfType(int handle, String kind) {
        synchronized (lock) {
            var kPtr = writeString(kind);
            try {
                var ptr = (int) documentBlocksOfType.apply(handle, kPtr)[0];
                return consumeString(ptr);
            } finally {
                if (kPtr != 0) dealloc.apply(kPtr, kind.getBytes(StandardCharsets.UTF_8).length + 1);
            }
        }
    }
}
